exchanges/orderbook/depth_test.go

```diff
@@ -712,1689 +504,409 @@ func TestPublish(t *testing.T) {
-	}
-
-	askT, bidT, err = depth.GetTranches(0)
-	if !errors.Is(err, nil) {
-		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
-	}
-
-	if len(askT) != 20 {
-		t.Fatalf("received: '%v' but expected: '%v'", len(askT), 20)
-	}
-
-	if len(bidT) != 20 {
-		t.Fatalf("received: '%v' but expected: '%v'", len(bidT), 20)
-	}
-
-	askT, bidT, err = depth.GetTranches(5)
-	if !errors.Is(err, nil) {
-		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
-	}
-
-	if len(askT) != 5 {
-		t.Fatalf("received: '%v' but expected: '%v'", len(askT), 5)
-	}
-
-	if len(bidT) != 5 {
-		t.Fatalf("received: '%v' but expected: '%v'", len(bidT), 5)
-	}
+	_ = depth.Invalidate(errors.New("invalid reasoning"))
+	return depth
 }
 
-func TestGetPair(t *testing.T) {
+func TestMovementMethods(t *testing.T) {
 	t.Parallel()
-	depth := NewDepth(id)
-
-	_, err := depth.GetPair()
-	if !errors.Is(err, currency.ErrCurrencyPairEmpty) {
-		t.Fatalf("received: '%v' but expected: '%v'", err, currency.ErrCurrencyPairEmpty)
-	}
-
-	expected := currency.NewPair(currency.BTC, currency.WABI)
-	depth.pair = expected
-
-	pair, err := depth.GetPair()
-	if !errors.Is(err, nil) {
-		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
-	}
-
-	if !pair.Equal(expected) {
-		t.Fatalf("received: '%v' but expected: '%v'", pair, expected)
-	}
-}
 
-func getInvalidDepth() *Depth {
-	depth := NewDepth(id)
-	_ = depth.Invalidate(errors.New("invalid reasoning"))
-	return depth
+	callMethod := func(i any, name string, args []any) (*Movement, error) {
+		m := reflect.ValueOf(i).MethodByName(name)
+		valueArgs := []reflect.Value{}
+		for _, i := range args {
+			valueArgs = append(valueArgs, reflect.ValueOf(i))
+		}
+		r := m.Call(valueArgs)
+		movement, ok := r[0].Interface().(*Movement)
+		assert.True(t, ok, "Should return an Movement type")
+		if err, ok := r[1].Interface().(error); ok {
+			return movement, err
+		}
+		return movement, nil
+	}
+
+	for _, tt := range movementTests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			depth := NewDepth(id)
+			methodName := strings.Split(tt.name, "_")[0]
+
+			_, err := callMethod(getInvalidDepth(), methodName, tt.tests[0].inputs)
+			assert.ErrorIsf(t, err, ErrOrderbookInvalid, "should error correctly with an invalid orderbook")
+
+			_, err = callMethod(depth, methodName, tt.tests[0].inputs)
+			assert.ErrorIs(t, err, errNoLiquidity, "should error correctly with no liquidity")
+
+			err = depth.LoadSnapshot(bid, ask, 0, time.Now(), true)
+			assert.NoError(t, err, "LoadSnapshot should not error")
+
+			for i, subT := range tt.tests {
+				move, err := callMethod(depth, methodName, subT.inputs)
+				assert.NoErrorf(t, err, "sub test %d should not error", i)
+				meta := reflect.Indirect(reflect.ValueOf(move))
+				metaExpect := reflect.Indirect(reflect.ValueOf(subT.expect))
+				for j := 0; j < metaExpect.NumField(); j++ {
+					field := meta.Field(j)
+					expect := metaExpect.Field(j)
+					// if !expect.IsZero() &&
```

Comment:
Damn, Leaked! Sorry!

I was testing what this would look like if I took out the clever checking and put in values for all fields.
Test failures were to 

Fixed: d81ff5984